geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/iterators/BinAggregatingScan.scala

```diff
@@ -15,96 +15,83 @@ import org.geotools.util.factory.Hints
+import org.locationtech.geomesa.index.iterators.BinAggregatingScan.ResultCallback
 import org.locationtech.geomesa.index.utils.bin.BinSorter
 import org.locationtech.geomesa.utils.bin.BinaryOutputEncoder.EncodingOptions
 import org.locationtech.geomesa.utils.bin.{BinaryOutputCallback, BinaryOutputEncoder}
 import org.locationtech.geomesa.utils.geotools.GeometryUtils
 import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
 import org.opengis.filter.Filter
 
-trait BinAggregatingScan extends AggregatingScan[ByteBufferResult] {
+trait BinAggregatingScan extends AggregatingScan[ResultCallback] {
 
   import BinAggregatingScan.Configuration._
 
-  var encoding: EncodingOptions = _
-  var encoder: BinaryOutputEncoder = _
-  var callback: ResultCallback = _
+  private var encoder: BinaryOutputEncoder = _
 
-  var binSize: Int = 16
-  var sort: Boolean = false
+  private var binSize: Int = 16
+  private var sort: Boolean = false
 
   // create the result object for the current scan
   override protected def initResult(
       sft: SimpleFeatureType,
       transform: Option[SimpleFeatureType],
-      options: Map[String, String]): ByteBufferResult = {
+      batchSize: Int,
+      options: Map[String, String]): ResultCallback = {
     val geom = options.get(GeomOpt).map(_.toInt).filter(_ != -1)
     val dtg = options.get(DateOpt).map(_.toInt).filter(_ != -1)
     val track = options.get(TrackOpt).map(_.toInt).filter(_ != -1)
     val label = options.get(LabelOpt).map(_.toInt).filter(_ != -1)
 
-    encoding = EncodingOptions(geom, dtg, track, label)
-    encoder = BinaryOutputEncoder(sft, encoding)
+    encoder = BinaryOutputEncoder(sft, EncodingOptions(geom, dtg, track, label))
 
     binSize = if (label.isEmpty) { 16 } else { 24 }
     sort = options(SortOpt).toBoolean
 
-    val batchSize = options(BatchSizeOpt).toInt * binSize
-
-    val buffer = ByteBuffer.wrap(Array.ofDim(batchSize)).order(ByteOrder.LITTLE_ENDIAN)
+    val buffer = ByteBuffer.wrap(Array.ofDim(batchSize * binSize)).order(ByteOrder.LITTLE_ENDIAN)
     val overflow = ByteBuffer.wrap(Array.ofDim(binSize * 16)).order(ByteOrder.LITTLE_ENDIAN)
 
-    callback = new ResultCallback(new ByteBufferResult(buffer, overflow))
-
-    callback.result
+    new ResultCallback(buffer, overflow)
   }
 
-  // add the feature to the current aggregated result
-  override protected def aggregateResult(sf: SimpleFeature, result: ByteBufferResult): Unit =
-    encoder.encode(sf, callback)
+  override protected def defaultBatchSize: Int =
+    throw new IllegalArgumentException("Batch scan is specified per scan")
 
-  override protected def notFull(result: ByteBufferResult): Boolean =
-    result.buffer.position < result.buffer.limit
+  // add the feature to the current aggregated result
+  override protected def aggregateResult(sf: SimpleFeature, result: ResultCallback): Int = {
+    val pos = result.position
+    encoder.encode(sf, result)
+    (result.position - pos) / binSize
```

Comment:
Do we ever encode more than one result at a time here?

Reply:
nope, this is called per feature

Reply:
If that's the case, will this calculation always be 1?

Reply:
ah, sorry, no. if it's a line string then it will encode to multiple 'bin' records so we still have to track it.